crates/core/src/rlp/structs.rs

```diff
@@ -0,0 +1,75 @@
+use super::{
+    decode::{decode_rlp_item, RLPDecode},
+    error::RLPDecodeError,
+};
+
+/// # Struct decoding helper.
+///
+/// Used to decode a struct from RLP format.
+/// The struct's fields must implement [`RLPDecode`].
+/// The struct is expected as a list, with its values being the fields
+/// in the order they are passed to [`Decoder::decode_field`].
+#[derive(Debug)]
+pub struct Decoder<'a> {
+    payload: &'a [u8],
+    remaining: &'a [u8],
+}
+
+impl<'a> Decoder<'a> {
+    pub fn new(buf: &'a [u8]) -> Result<Self, RLPDecodeError> {
+        match decode_rlp_item(buf)? {
+            (true, payload, remaining) => Ok(Self { payload, remaining }),
+            (false, _, _) => Err(RLPDecodeError::UnexpectedList),
```

Comment:
I think here we should use something else, because `RLPDecodeError::UnexpectedList` is meant to be used when you encounter a list but you wasn't expecting one

Reply:
Good catch!